File: codingjs/countHi.js
function countHi(str: string): number {
  if (str.length < 3) return str === 'hi' ? 1 : 0;
  if (str.substring(0, 2) === 'hi') {
    return 1 + countHi(str.substring(2));
  } else {
    return 0 + countHi(str.substring(1));
  }
}

File: codingbat/countHi.java
public int countHi(String str) {
  if (str.length() < 2) return 0;
  int count = 0;
  for (int i = 0; i < str.length() - 1; i++) {
    if (str.substring(i, i + 2).equals("hi")) {
      count++;
    }
  }
  return count;
}

